### ISSUE TITLE
Prototype a new CLI bmhvm

### ISSUE THREAD

**Issue by Rozzii** [metal3-io/metal3-dev-env#1370]
This is an advanced task. Take it only if you are already comfortable with golang, libvirt and containers.

We currently use bash scripts to [set up sushy-tools](https://github.com/metal3-io/baremetal-operator/blob/06f7d39530f943791de6aab12c3791893db873e9/hack/ci-e2e.sh#L80-L87) and [create VMs to be used as BareMetalHosts](https://github.com/metal3-io/baremetal-operator/blob/06f7d39530f943791de6aab12c3791893db873e9/hack/ci-e2e.sh#L94-L95). This works, but it is not ideal. It is inflexible, hard to reuse between CI and development, and even harder to reuse from other repositories like CAPM3.

The goal of this task is to prototype a golang CLI (call it for example `bmhvm`) that can do the same thing these scripts do today. It should also be usable as a go library that can be imported in CAPM3 for example. The tool should be able to set up sushy-tools on top of either podman or docker. It should be able to create a libvirt network and create VMs in this network.

You will need to work with the libvirt golang library: https://pkg.go.dev/libvirt.org/go/libvirt

Here are some suggestions for how it should work.

- `bmhvm init` should create a libvirt network and run sushy-tools
- `bmhvm create vm` should create a VM
- `bmhvm generate bmh` should generate a BareMetalHost manifest for a VM

Note that you can reuse some parts [here](https://github.com/metal3-io/baremetal-operator/tree/main/cmd).

**Comment by dtantsur** [comment]
@Rozzii would it help if we put that in sushy-tools instead? I'm a bit worried about potential duplication: we already have many places where the same logic is used.

**Comment by Rozzii** [comment]
@dtantsur yeah I think we can go into that direction, and as you have mentioned let's first specify the exact scope of this and figure out the configuration format.
/triage accepted

**Comment by Rozzii** [comment]
/remove-lifecycle frozen
We had a discussion with @mquhuy and realized that he has done much of he hard work when in terms of implementing go based libvirt tooling here https://github.com/metal3-io/baremetal-operator/tree/main/test/createVM.

After some discussion we have decided that this VM tooling would be used across the project and the best place for it would be in project-infra. CAPM3, IPAM and BMO would use it as a GO package while dev-env would use it as a cli tool.

**Comment by tuminoid** [comment]
> I'm not sure project-infra is the best place. Yes it will be used across projects, but I do think it belongs with BMO. We should publish the CLI as a build artifact binary that dev-env can use. In CAPM3 I think we should use it as a go module directly, and this is the reason I want it in BMO. CAPM3 depends on BMO already and this is a natural dependency. Depending on project-infra as a module feels wrong to me.

I agree that CAPM3 or BMO should not depend on go package from project-infra. I'm not exactly loving the idea of having VM tools in BMO either, but it is more closely related to BMO than anything else.
